core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -739,6 +813,61 @@ public GroupByIdBlock getResult()
         }
     }
 
+    private class GetLowCardinalityDictionaryGroupIdsWork
+            extends LowCardinalityDictionaryWork<GroupByIdBlock>
+    {
+        private final long[] groupIds;
+        private boolean finished;
+
+        public GetLowCardinalityDictionaryGroupIdsWork(Page page)
+        {
+            super(page);
+            groupIds = new long[page.getPositionCount()];
+        }
+
+        @Override
+        public boolean process()
+        {
+            int positionCount = page.getPositionCount();
+            checkState(lastPosition < positionCount, "position count out of bound");
+
+            int[] samplePositions = new int[maxCardinality];
+            Arrays.fill(samplePositions, -1);
+            short[] positionToCombination = calculatePositions(positionCount, samplePositions);
+
+            // needRehash() == false indicates we have reached capacity boundary and a rehash is needed.
+            // We can only proceed if tryRehash() successfully did a rehash.
+            if (needRehash() && !tryRehash()) {
+                return false;
+            }
+
+            // putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.
+            // Therefore needRehash will not generally return true even if we have just crossed the capacity boundary.
+            for (int i = 0; i < maxCardinality; i++) {
+                if (needRehash()) {
+                    return false;
+                }
+                if (samplePositions[i] != -1) {
+                    samplePositions[i] = putIfAbsent(samplePositions[i], page);
```

Comment:
use a new array instead of reusing `samplePositions`. It probably doesn't matter for performance (might be faster since we don't change cached row in CPU)